ssl/ech/ech_internal.c

```diff
@@ -160,4 +227,1223 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+    /* padding is after the inner client hello has been encoded */
+    length_with_snipadding = innersnipadding + s->ext.ech.encoded_innerch_len;
+    length_of_padding = 31 - ((length_with_snipadding - 1) % 32);
+    length_with_padding = s->ext.ech.encoded_innerch_len
+        + length_of_padding + innersnipadding;
+    /*
+     * Finally - make sure final result is longer than padding target
+     * and a multiple of our padding increment.
+     * TODO(ECH): This is a local addition - we might take it out if
+     * it makes us stick out; or if we take out the above more (uselessly:-)
+     * complicated scheme, we may only need this in the end.
+     */
+    if (length_with_padding % OSSL_ECH_PADDING_INCREMENT)
+        length_with_padding += OSSL_ECH_PADDING_INCREMENT
+            - (length_with_padding % OSSL_ECH_PADDING_INCREMENT);
+    while (length_with_padding < OSSL_ECH_PADDING_TARGET)
+        length_with_padding += OSSL_ECH_PADDING_INCREMENT;
+    clear_len = length_with_padding;
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "EAAE: padding: mnl: %zu, lws: %d "
+                   "lop: %d, lwp: %d, clear_len: %zu, orig: %zu\n",
+                   mnl, length_with_snipadding, length_of_padding,
+                   length_with_padding, clear_len,
+                   s->ext.ech.encoded_innerch_len);
+    } OSSL_TRACE_END(TLS);
+    return clear_len;
+}
+
+/*
+ * Calculate AAD and do ECH encryption
+ * pkt is the packet to send
+ * return 1 for success, other otherwise
+ *
+ * 1. Make up the AAD: the encoded outer, with ECH ciphertext octets zero'd
+ * 2. Do the encryption
+ * 3. Put the ECH back into the encoding
+ * 4. Encode the outer (again!)
+ */
+int ech_aad_and_encrypt(SSL_CONNECTION *s, WPACKET *pkt)
+{
+    int rv = 0, hpke_mode = OSSL_HPKE_MODE_BASE;
+    OSSL_ECHSTORE_ENTRY *ee = NULL;
+    OSSL_HPKE_SUITE hpke_suite = OSSL_HPKE_SUITE_DEFAULT;
+    unsigned char config_id_to_use = 0x00, info[SSL3_RT_MAX_PLAIN_LENGTH];
+    unsigned char *clear = NULL, *cipher = NULL, *aad = NULL, *mypub = NULL;
+    size_t cipherlen = 0, aad_len = 0, lenclen = 0, mypub_len = 0;
+    size_t info_len = SSL3_RT_MAX_PLAIN_LENGTH, clear_len = 0;
+
+    if (s == NULL)
+        return 0;
+    if (s->ext.ech.es == NULL || s->ext.ech.es->entries == NULL
+        || pkt == NULL || s->ssl.ctx == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    rv = ech_pick_matching_cfg(s, &ee, &hpke_suite);
+    if (rv != 1 || ee == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    s->ext.ech.attempted_type = ee->version;
+    OSSL_TRACE_BEGIN(TLS) {
+        BIO_printf(trc_out, "EAAE: selected: version: %4x, config %2x\n",
+                   ee->version, ee->config_id);
+    } OSSL_TRACE_END(TLS);
+    config_id_to_use = ee->config_id; /* if requested, use a random config_id instead */
+    if (s->ssl.ctx->options & SSL_OP_ECH_IGNORE_CID
```

Comment:
When you are using bit-operator, you must parenthesis these operation.
This should be write it like this : 
````C
if ((s->ssl.ctx->options & SSL_OP_ECH_IGNORE_CID) != 0
        || (s->options & SSL_OP_ECH_IGNORE_CID) != 0) {
```

Reply:
is that true? the `SSL_OP_*` values are single bit masks so I don't believe added parentheses are needed, however I see that pattern is followed at least in `ssl/ssl_lib.c` so no harm to do it...

Reply:
It is about "C operators precedence".
Logical operations are evaluated before bit-operations.